import os
import numpy as np
import sys
import ytil


'''
A deep MSA is usually preferable but not always better than a
shallow MSA (see the examples provided in ref. 3). In this work, 5 alternative
alignments are generated for each target. The first 4 are generated independently
by searching the Uniclust30 database (version 2018_08) with
HHblits (version 3.0.3) (28) with default parameters at 4 different e-value
cutoffs: 1e−40, 1e−10, 1e−3, and 1. The last alignment was generated by
several rounds of iterative HHblits searches with gradually relaxed e-value
cutoffs (1e−80, 1e−70,. . ., 1e−10, 1e−8, 1e−6, and 1e−4), followed by the
hmmsearch (version 3.1b2) (29) against the metagenome sequence database
(20) in case not enough sequences were collected at previous steps. The
metagenome database includes about 7 billion protein sequences from the
following resources: 1) JGI Metagenomes (7,835 sets), Metatranscriptomes
(2,623 sets), and Eukaryotes (891 genomes); 2) UniRef100; 3) NCBI TSA (2,616
        sets); and 4) genomes manually collected from various genomic centers and
online depositories (2,815 genomes). 

To avoid attracting distant homologs
at early stages and making alignment unnecessarily deep, the search was
stopped whenever either of the 2 criteria were met: at least 2,000 sequences
with 75% coverage or 5,000 sequences with 50% coverage (both at 90% sequence
        identity cutoff) were collected.
'''

def seq__npy(lfp, name=None):
    seq = np.load( lfp, allow_pickle=True).item()['SEQ']
    return ['>'+name, seq]


'''
loaddir = '/raid0/ychnh/ca_meta'
savedir = '/raid0/ychnh/seq'
for meta in os.listdir(loaddir):
    name = meta[:-4]
    lfp = os.path.join(loaddir, meta)
    sfp = os.path.join(savedir, name+'.seq')
    ytil.file__list( seq__npy(lfp, name), sfp)
'''

# TODO: Have a fixed list to iterate through

def a3m__seq(seq_fp, a3m_fp, database, e=.001, n=1, cpu=10):
    os.system('hhblits -o temp.hhr -cpu '+str(cpu)+' -i '+seq_fp+' -d '+database+' -oa3m '+a3m_fp+' -n '+str(n)+ ' -e '+str(e))

database = '/raid0/uniclust/uniclust30_2018_08/uniclust30_2018_08'
loaddir = '/raid0/ychnh/seq'
savedir = '/raid0/ychnh/a3m'

for seqfp in os.listdir(loaddir):
    name = seqfp[:-4]
    in_fp = os.path.join(loaddir, seqfp)
    out_fp = os.path.join(savedir, name+'.a3m')
    a3m__seq(in_fp, out_fp, database)
 
#a3m__seq('PAR.seq', 'PAR.a3m', database)
#os.system('hhblits -cpu 1 -i 3WTG.seq -d ~/hh-suit/databases/uniclust30_2018_08/uniclust30_2018_08 -oa3m align.a3m -ohhm align.hhm -n 3')
